File: QuantConnect/Strategy_23.cs


# Strategy_TrendFollowing.py

from QuantConnect.Data import *
from QuantConnect.Indicators import *
from QuantConnect.Algorithm import *
from QuantConnect.Algorithm.Framework import *

class TrendFollowing(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2015, 1, 1)  # Set Start Date
        self.SetEndDate(2020, 12, 31)  # Set End Date
        self.SetCash(100000)  # Set Strategy Cash

        self.symbol = self.AddEquity("SPY", Resolution.Daily).Symbol
        self.supertrend = self.SuperTrend(self.symbol, 10)
        self.atr = self.ATR(self.symbol, 14)

        # Set up trailing stop
        self.trailing_stop = None

        # Buy/Sell signals
        self.buy_triggered = False
        self.sell_triggered = False

    def OnData(self, data):
        if not (self.supertrend.IsReady and data.ContainsKey(self.symbol)):
            return

        price = data[self.symbol].Close

        if price > self.supertrend.UpperBand.Current.Value and not self.buy_triggered:
            # Buy signal - price crosses upper band
            stop_price = price - (self.atr.Current.Value * 1)
            limit_price = None  # No limit order

            # Place order and set up trailing stop
            order_ticket = self.MarketOrder(self.symbol, 100)
            if order_ticket.Status == OrderStatus.Filled:
                stop_price_str = '{:.2f}'.format(stop_price)
                trailing_stop_info = TrailingStopRiskManagementModel.ExitAtStopPrice(order_ticket.OrderId, Decimal(stop_price_str))
                self.trailing_stop = QuantityTrailingStopRiskManagementModel(trailing_stop_info)

                # Reset buy/sell triggers
                self.buy_triggered = True
                self.sell_triggered = False

        elif price < self.supertrend.LowerBand.Current.Value and not sell_triggered:
            # Sell signal - price crosses lower band or stop loss triggered by trailing stop
            stop_price = None  # No stop loss order since trailing stop is in place
            limit_price = None  # No limit order

            # Place order and cancel existing trailing stop (if any)
            order_ticket = self.MarketOrder(self.symbol, -100)
            if order_ticket.Status == OrderStatus.Filled:
                if (self.trailing_stop is not None):
                    if not (self.trailing_stop.Cancel(order_ticket.OrderId)):
                        return

                # Reset buy/sell triggers
                sell_triggered=True 
                buy_triggered=False

File: QuantConnect/Strategy_23.py


# Strategy_TrendFollowing.py

from SuperTrend import SuperTrend
from System import *
from QuantConnect import *
from QuantConnect.Data.Market import TradeBar
from QuantConnect.Orders import OrderStatus
from QuantConnect.Orders.Fees import ConstantFeeModel
from QuantConnect.Orders.Fills import ImmediateFillModel
from QuantConnect.Algorithm.Framework.Alphas.Trend import *
from QuantConnect.Algorithm.Framework.Execution.ImmediateExecutionModel import ImmediateExecutionModel
from QuantConnect.Algorithm.Framework.Portfolio.EqualWeightingPortfolioConstructionModel import EqualWeightingPortfolioConstructionModel


class TrendFollowing(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2020, 12, 31)
        self.SetCash(100000)
        self.SetBrokerageModel(BrokerageName.InteractiveBrokersBrokerage,
                               AccountType.Margin)

        # Define the securities to be traded
        self.tickers = ['SPY']
        for ticker in self.tickers:
            equity = self.AddEquity(ticker, Resolution.Daily)
            equity.SetDataNormalizationMode(DataNormalizationMode.Raw)

        # Add the SuperTrend indicator to the chart
        indicatorPeriod = 10
        atrPeriod = 14
        multiplier = 2.0
        superTrendIndicator = SuperTrend(indicatorPeriod, atrPeriod, multiplier)
        self.PlotIndicator("SuperTrend", superTrendIndicator)

        # Use the SuperTrend as the primary indicator for trend following strategy
        superTrendCrossAboveThreshold = IndicatorCrossAbove(superTrendIndicator.SuperTrendUpperBand,
                                                            equity.Price,
                                                            timedelta(days=1))
        superTrendCrossBelowThreshold = IndicatorCrossBelow(superTrendIndicator.SuperTrendLowerBand,
                                                             equity.Price,
                                                             timedelta(days=1))
        
         # Define Alpha Model to generate buy signals when price crosses above upper band and sell signals when price crosses below lower band.
         alpha_model = InsightManager(CompositeAlphaModel(
                TrendFollowingAlphaModel(
                    resolution=Resolution.Daily,
                    super_trend_cross_above_threshold=super_trend_cross_above_threshold,
                    super_trend_cross_below_threshold=super_trend_cross_below_threshold),
                NullAlphaModel()))

         # Define Portfolio Construction Model as Equal Weighting Portfolio Construction Model.
         portfolio_construction_model = EqualWeightingPortfolioConstructionModel()

         # Define Execution Model as Immediate Execution Model.
         execution_model = ImmediateExecutionModel()

         # Set up fee and fill models.
         fee_model = ConstantFeeModel(0)
         fill_model = ImmediateFillModel()

         # Set up algorithm with defined models.
         self.SetUniverseSelection(self.UniverseSelection.None)
         self.SetAlpha(alpha_model)
         self.SetPortfolioConstruction(portfolio_construction_model)
         self.SetExecution(execution_model)
         self.SetRiskManagement(TrailingStopRiskManagementModel(1.0))  # Set stop criteria as 1 x ATR.

    def OnData(self, data):
            pass